units/pod_definition_creation.go

```diff
@@ -0,0 +1,206 @@
+package units
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/evergreen-ci/cocoa"
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/cloud"
+	"github.com/evergreen-ci/evergreen/model/pod"
+	"github.com/evergreen-ci/evergreen/model/pod/definition"
+	"github.com/evergreen-ci/utility"
+	"github.com/mongodb/amboy"
+	"github.com/mongodb/amboy/job"
+	"github.com/mongodb/amboy/registry"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+)
+
+const (
+	podDefinitionCreationJobName     = "pod-definition-creation"
+	podDefinitionCreationMaxAttempts = 15
+)
+
+func init() {
+	registry.AddJobType(podDefinitionCreationJobName, func() amboy.Job {
+		return makePodDefinitionCreationJob()
+	})
+}
+
+type podDefinitionCreationJob struct {
+	job.Base      `bson:"metadata" json:"metadata" yaml:"metadata"`
+	ContainerOpts pod.TaskContainerCreationOptions `bson:"container_opts" json:"container_opts" yaml:"container_opts"`
+	Family        string                           `bson:"family" json:"family" yaml:"family"`
+
+	ecsClient        cocoa.ECSClient
+	ecsPodDefManager cocoa.ECSPodDefinitionManager
+	smClient         cocoa.SecretsManagerClient
+	vault            cocoa.Vault
+	env              evergreen.Environment
+	settings         evergreen.Settings
+}
+
+func makePodDefinitionCreationJob() *podDefinitionCreationJob {
+	j := &podDefinitionCreationJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    podDefinitionCreationJobName,
+				Version: 0,
+			},
+		},
+	}
+	return j
+}
+
+// NewPodDefinitionCreationJob creates a job that creates a pod definition in
+// preparation for running a pod.
+func NewPodDefinitionCreationJob(ecsConf evergreen.ECSConfig, opts pod.TaskContainerCreationOptions, id string) amboy.Job {
+	j := makePodDefinitionCreationJob()
+	j.ContainerOpts = opts
+	j.Family = j.ContainerOpts.GetFamily(ecsConf)
+	j.SetID(fmt.Sprintf("%s.%s.%s", podDefinitionCreationJobName, j.Family, id))
+	j.SetScopes([]string{fmt.Sprintf("%s.%s", podDefinitionCreationJobName, j.Family)})
+	j.SetEnqueueAllScopes(true)
+	j.UpdateRetryInfo(amboy.JobRetryOptions{
+		Retryable:   utility.TruePtr(),
+		MaxAttempts: utility.ToIntPtr(podDefinitionCreationMaxAttempts),
+		WaitUntil:   utility.ToTimeDurationPtr(10 * time.Second),
+	})
+
+	return j
+}
+
+func (j *podDefinitionCreationJob) Run(ctx context.Context) {
+	defer j.MarkComplete()
+
+	defer func() {
```

Comment:
Can we combine these two deferred functions?

Reply:
Done.